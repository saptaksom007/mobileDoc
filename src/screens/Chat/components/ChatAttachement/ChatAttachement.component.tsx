import * as React from 'react'
import axios from 'axios'
import {
  View,
  TouchableOpacity,
  InteractionManager,
  TouchableWithoutFeedback,
  Image as RNImage,
} from 'react-native'
import { SafeImage as Image } from 'react-native-safe-image'

import { connect } from 'react-redux'

import PhotoViewerActions from 'screens/PhotoViewer/PhotoViewer.actions'
import PdfReaderActions from 'screens/PdfReader/PdfReader.actions'
import { Loader } from 'components/Loader/Loader.component'
import { Env } from 'env'
import { Layout } from 'constants/Layout'

import styles from './ChatAttachement.styles'
import { ChatMessage, MediaResource } from 'api/types'
import { Action } from 'redux'
import { PreviewType } from 'utilities/commonViewerReducer'
import { dispatchAsync } from 'utilities/dispatch.helpers'

const isPdf = (mimeType?: string) => mimeType === 'application/pdf'

const FallbackComponent = React.memo(() => (
  <View style={styles.fallback}>
    <Loader containerProps={{ style: styles.loaderContainer }} />
  </View>
))

interface Props {
  message: ChatMessage
  dispatch(action: Action): void
  position: 'left' | 'right'
  pdfreader: {
    previews: PreviewType
  }
  photoviewer: {
    previews: PreviewType
  }
  canFetchPreview: boolean
}

interface State {
  width: number
  height: number
  previewWidth: number
  previewHeight: number
  isPending: boolean
  embedFailed?: boolean
}

export class ChatAttachement extends React.Component<Props, State> {
  state: State = {
    width: 0,
    height: 0,
    previewWidth: Layout.window.width,
    previewHeight: Layout.window.height,
    embedFailed: false,
    isPending: false,
  }

  constructor(props: Props) {
    super(props)
    const { canFetchPreview } = this.props
    if (canFetchPreview) {
      this.fetchPreview()
    }
  }

  fetchPreview = () => {
    const { message, dispatch, pdfreader, photoviewer } = this.props
    const mediaId = message.mediaResource?.uuid!

    if (message.isEmbedImage && message.image) {
      axios({
        url: message.image,
        method: 'HEAD',
      })
        .then(response => {
          const contentType =
            response.headers['Content-Type'] || response.headers['content-type']
          if (contentType?.startsWith('image')) {
            RNImage.prefetch(message.image!)
            this.setState({ embedFailed: false })
          } else {
            this.setState({ embedFailed: true })
          }
        })
        .catch(() => {
          this.setState({ embedFailed: true })
        })
      return
    }

    if (isPdf(message.mediaResource?.mimeType!)) {
      if (
        mediaId &&
        (!pdfreader.previews[mediaId] ||
          (pdfreader.previews[mediaId] && !pdfreader.previews[mediaId].dataUrl))
      ) {
        dispatch(
          PdfReaderActions.pdfReaderPreviewRequested(message.mediaResource),
        )
      }
    } else {
      if (
        mediaId &&
        (!photoviewer.previews[mediaId] ||
          (photoviewer.previews[mediaId] &&
            !photoviewer.previews[mediaId].dataUrl))
      ) {
        dispatch(
          PhotoViewerActions.photoViewerPreviewRequested(message.mediaResource),
        )
      }
    }
  }

  shouldComponentUpdate(nextProps: Props) {
    const { message, pdfreader, photoviewer, canFetchPreview } = this.props

    if (canFetchPreview !== nextProps.canFetchPreview) {
      return true
    }

    const mediaId = message.mediaResource?.uuid!
    if (message.mediaResource && isPdf(message.mediaResource?.mimeType!)) {
      if (
        !nextProps?.pdfreader?.previews[mediaId] ||
        !pdfreader.previews[mediaId]
      ) {
        return true
      }
      return (
        (pdfreader.previews[mediaId] &&
          nextProps?.pdfreader?.previews[mediaId] &&
          nextProps?.pdfreader?.previews[mediaId].loading) ||
        pdfreader.previews[mediaId].dataUrl !==
          nextProps.pdfreader.previews[mediaId].dataUrl
      )
    } else {
      if (
        !nextProps?.photoviewer?.previews[mediaId] ||
        !photoviewer.previews[mediaId]
      ) {
        return true
      }
      return (
        (photoviewer.previews[mediaId] &&
          nextProps?.photoviewer?.previews[mediaId] &&
          nextProps?.photoviewer?.previews[mediaId].loading) ||
        photoviewer.previews[mediaId].dataUrl !==
          nextProps.photoviewer.previews[mediaId].dataUrl
      )
    }
  }

  componentDidUpdate(prevProps: Props) {
    if (
      prevProps.canFetchPreview === false &&
      this.props.canFetchPreview === true
    ) {
      this.fetchPreview()
    }
  }

  getUri() {
    const {
      message: { mediaResource, isEmbedImage, image },
      pdfreader: { previews: pdfPreview },
      photoviewer: { previews: photoPreview },
    } = this.props
    if (mediaResource && isPdf(mediaResource?.mimeType!)) {
      return (
        pdfPreview &&
        pdfPreview[mediaResource?.uuid!] &&
        pdfPreview[mediaResource?.uuid!].dataUrl
      )
    }

    if (
      photoPreview &&
      photoPreview[mediaResource?.uuid!] &&
      photoPreview[mediaResource?.uuid!].dataUrl &&
      !isEmbedImage
    ) {
      return photoPreview[mediaResource?.uuid!].dataUrl
    }

    return image
  }

  openImage = async () => {
    const { dispatch, message } = this.props
    const { mediaResource, image } = message

    this.setState({ isPending: true })
    if (!message.isEmbedImage) {
      await dispatchAsync(
        dispatch,
        PhotoViewerActions.navigateToPhotoViewer(
          mediaResource as MediaResource,
        ),
      )
    } else {
      await dispatchAsync(
        dispatch,
        PhotoViewerActions.navigateToPhotoViewer(image as string),
      )
    }

    this.setState({ isPending: false })
  }

  isLoadingPreview = () => {
    const { pdfreader, photoviewer, message } = this.props
    const { mediaResource } = message

    if (message.isEmbedImage) {
      return false
    }

    if (!mediaResource || !pdfreader.previews || !photoviewer.previews) {
      return true
    }

    if (mediaResource && isPdf(mediaResource!.mimeType!)) {
      return (
        pdfreader.previews &&
        pdfreader.previews[mediaResource!.uuid!] &&
        pdfreader.previews[mediaResource!.uuid!].loading
      )
    }

    return (
      photoviewer.previews &&
      photoviewer.previews[mediaResource!.uuid!] &&
      photoviewer.previews[mediaResource!.uuid!].loading
    )
  }

  renderPreview() {
    const uri = this.getUri()
    if (uri) {
      return (
        <Image
          source={{ uri }}
          style={styles.image}
          blurRadius={1}
          fallbackComponent={<FallbackComponent />}
        />
      )
    }
    return <FallbackComponent />
  }

  renderImage() {
    const { isPending } = this.state
    return (
      <TouchableOpacity disabled={isPending} onPress={this.openImage}>
        {this.renderPreview()}
      </TouchableOpacity>
    )
  }

  handleGoToPdf = () => {
    const { message, dispatch } = this.props
    this.setState({ isPending: true }, () => {
      dispatch(PdfReaderActions.pdfReaderPush(message.mediaResource))
    })
    InteractionManager.runAfterInteractions(() => {
      this.setState({ isPending: false })
    })
  }
  renderPdf() {
    const { isPending } = this.state
    return (
      <TouchableOpacity disabled={isPending} onPress={this.handleGoToPdf}>
        <Image
          source={{
            uri: `${Env.api.base}/messaging/pdf.jpeg`,
            cache: 'force-cache',
          }}
          style={styles.image}
          blurRadius={1}
        />
      </TouchableOpacity>
    )
  }

  handleGoToDetail = () => {
    const { dispatch, message } = this.props
    const { mediaResource } = message
    const isLoading = this.isLoadingPreview()
    if (!isLoading) {
      if (isPdf(mediaResource?.mimeType!)) {
        dispatch(PdfReaderActions.pdfReaderPush(mediaResource))
      } else {
        this.openImage()
      }
    }
  }

  render() {
    const { message } = this.props
    const { mediaResource, isEmbedImage } = message
    if (
      (!mediaResource && !isEmbedImage) ||
      this.state.embedFailed ||
      message.isEmbedVideo
    ) {
      return null
    }
    const content = !isPdf(mediaResource && mediaResource!.mimeType!)
      ? this.renderImage()
      : this.renderPdf()
    return (
      <View style={styles.container}>
        <TouchableWithoutFeedback onPress={this.handleGoToDetail}>
          <View style={styles.overlay} />
        </TouchableWithoutFeedback>
        {content}
      </View>
    )
  }
}

function select({
  pdfreader,
  photoviewer,
  loadingStack: {
    pendingActions: { PHOTOVIEWER_PREVIEW },
  },
}: any) {
  return {
    pdfreader,
    photoviewer,
    canFetchPreview:
      PHOTOVIEWER_PREVIEW <= 10 || PHOTOVIEWER_PREVIEW === undefined,
  }
}

export default connect(select)(ChatAttachement)
