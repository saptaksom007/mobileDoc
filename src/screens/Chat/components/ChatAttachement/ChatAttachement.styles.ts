import { StyleSheet } from 'react-native'

const borderRadius = 13

export default StyleSheet.create({
  container: {
    flex: 1,
    padding: 3,
    borderRadius,
    minWidth: 200,
    maxHeight: 100,
  },
  loader: {
    zIndex: 10,
  },
  loaderContainer: {
    backgroundColor: 'transparent',
  },
  image: {
    minWidth: 200,
    height: 100,
    padding: 7,
    borderRadius,
    resizeMode: 'cover',
    width: '100%',
  },
  contentContainerStyle: {
    flex: 1,
  },
  imageActive: {
    flex: 1,
    resizeMode: 'contain',
  },
  transformImageActive: {
    flex: 1,
  },
  overlay: {
    position: 'absolute',
    height: 100,
    width: '100%',
    borderRadius,
    left: 3,
    top: 3,
    zIndex: 3,
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'rgba(255,255,255,0.4)',
  },
  fallback: {
    minWidth: 200,
    height: 100,
    padding: 7,
    width: '100%',
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'rgba(255,255,255,0.4)',
    borderRadius,
  },
})
