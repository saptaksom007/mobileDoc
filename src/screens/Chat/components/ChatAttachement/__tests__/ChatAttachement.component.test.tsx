import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import { ChatAttachement } from '../ChatAttachement.component'

it('<Chat/> render correctly', () => {
  const tree = renderer
    .create(
      <ChatAttachement
        position='right'
        dispatch={() => {}}
        pdfreader={{
          previews: {},
        }}
        photoviewer={{
          previews: {},
        }}
        canFetchPreview
        message={{
          _id: '123',
          user: {
            _id: '456',
          },
          createdAt: new Date(),
          text: '',
          mediaResource: {
            previewUrl: 'http://via.placeholder.com/350x150',
            originalUrl: 'http://via.placeholder.com/350x150',
          },
          meta: {
            originalSize: 123,
          },
        }}
      />,
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
