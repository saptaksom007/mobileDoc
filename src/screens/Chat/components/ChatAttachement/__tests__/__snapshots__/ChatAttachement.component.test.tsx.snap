// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chat/> render correctly 1`] = `
<View
  style={
    Object {
      "borderRadius": 13,
      "flex": 1,
      "maxHeight": 100,
      "minWidth": 200,
      "padding": 3,
    }
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(255,255,255,0.4)",
        "borderRadius": 13,
        "flex": 1,
        "height": 100,
        "justifyContent": "center",
        "left": 3,
        "position": "absolute",
        "top": 3,
        "width": "100%",
        "zIndex": 3,
      }
    }
  />
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(255,255,255,0.4)",
          "borderRadius": 13,
          "height": 100,
          "justifyContent": "center",
          "minWidth": 200,
          "padding": 7,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#fff",
              "flex": 1,
              "justifyContent": "center",
            },
            Object {
              "backgroundColor": "transparent",
            },
          ]
        }
      >
        <ActivityIndicator
          animating={true}
          color="#888"
          hidesWhenStopped={true}
          size="large"
        />
        <Text
          allowFontScaling={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "fontFamily": "TitilliumWeb-Regular",
                "fontSize": 17.5,
                "lineHeight": 22,
                "overflow": "visible",
              },
              Object {
                "color": "rgba(27, 106, 158, 0.85)",
                "fontSize": 15,
                "lineHeight": 18,
              },
              undefined,
            ]
          }
        >
          
        </Text>
      </View>
    </View>
  </View>
</View>
`;
